xiilib/flask/charm.py

```diff
@@ -1,38 +1,32 @@
 #!/usr/bin/env python3
+# pylint: disable=duplicate-code
+
 # Copyright 2024 Canonical Ltd.
 # See LICENSE file for licensing details.
 
 """Flask Charm service."""
 
 import logging
 import pathlib
-import textwrap
 import typing
 
 import ops
-from charms.data_platform_libs.v0.data_interfaces import DatabaseRequiresEvent
 from charms.traefik_k8s.v2.ingress import IngressPerAppRequirer
 
-from xiilib.database_migration import DatabaseMigration, DatabaseMigrationStatus
-from xiilib.databases import Databases, get_uris, make_database_requirers
+from xiilib._gunicorn.charm import CharmMixin
+from xiilib._gunicorn.observability import Observability
+from xiilib._gunicorn.secret_storage import GunicornSecretStorage
+from xiilib._gunicorn.webserver import GunicornWebserver
+from xiilib._gunicorn.wsgi_app import WsgiApp
+from xiilib.database_migration import DatabaseMigration
+from xiilib.databases import Databases, make_database_requirers
 from xiilib.exceptions import CharmConfigInvalidError
 from xiilib.flask.charm_state import CharmState
-from xiilib.flask.constants import (
-    FLASK_APP_DIR,
-    FLASK_BASE_DIR,
-    FLASK_CONTAINER_NAME,
-    FLASK_SERVICE_NAME,
-    FLASK_STATE_DIR,
-)
-from xiilib.flask.flask_app import FlaskApp
-from xiilib.flask.secret_storage import FlaskSecretStorage
-from xiilib.observability import Observability
-from xiilib.webserver import GunicornWebserver
 
 logger = logging.getLogger(__name__)
 
 
-class Charm(ops.CharmBase):  # pylint: disable=too-many-instance-attributes
+class Charm(ops.CharmBase, CharmMixin):  # pylint: disable=too-many-instance-attributes
```

Comment:
From memory, the discussion was that there should be a super class which handles all the different frameworks, and specific frameworks should inherit from that and make the required changes. What is the reason for changing that to a mixin please?

Reply:
Yes, and the superclass doesn't have its own constructor, relying on the subclass to instantiate framework-specific attributes such as different web servers and application classes. That's usually called mixin.

Reply:
It seems that a lot of what is done in `__init__` of flask is not specific to flask or could take the flask specific items as an argument, e.g. the secret storage key, couldn't that be done in the parent constructor if it was a normal parent class rather than a mixin?

Reply:
The major difference between `flask` and `django` and other frameworks is that they are using a different "charm state" class. I guess we can put the constructor in the parent class if we do a dependency inversion for charm state, is this what you would prefer?

Reply:
What would dependency inversion look like here?

One of the approaches could be that `CharmState` is constructed on the parent and is generic on the application config (represented by `FlaskConfig` in the case of flask) and takes that an instance of `FlaskConfig` or `DjangoConfig` etc as an argument. Would that work? The `<framework>Config` can be constructed on the framework specific class

Reply:
```python
class AppConfig(typing.Protocol):

    def environment() -> ...:


class CharmState:
    app_config: AppConfig

     def __init__(..., app_config: AppConfig):
         ...
```

Reply:
Can also be a pydantic base model

Reply:
Hi, I tried with your request and it seems to be impossible to implement, should the super class constructor or sub class constructor call the constructor for `ops.CharmBase`?

Reply:
I was thinking the super class? What is the issue please?

Reply:
The `app_config` need an `ops.CharmBase` to create, but `app_config` need to be passed to the super class's constructor, and `ops.CharmBase` is only available after super class's constructor returns? This is not possible logically in this universe I think.

Reply:
I guess it would be simpler to put the constructor of `ops.CharmBase` in the sub class, but this way I basically need to call a lot of `type.cast` in the super class's constructor because there's no way for Python to express a class is A and B (intersection types)

Reply:
Yeh I see. I think the options are:

1. A `<framework>Charm` needs to provide a method for constructing the `<framework>Config` like `make_config`
2. Instead of passing the instance, the class for `<framework>Config` can be passed to the parent and it can be constructed there
3. The mixin approach

Reply:
From typing's point of view, using a abstract method (1) might be the simplest to do

Reply:
Wouldn't both cases need a protocol or something like that since the `make_config` function would need to be typed in the parent class as well for what it returns and the parent `CharmState` needs to have a protocol or something like that for the `<framework>Config` attribute? I don't have a strong preference between (1) and (2)

Reply:
In this use case, passing a class as a function argument is essentially passing a factory function, which is a little bit complicated consider wether you should use `typing.Callable` or  `typing.Type`/`type`

Reply:
I'm happy to go with (1)

Reply:
I discussed this with David, and I'm good to go with (1). I think generally the approach of an abstract class or a protocol is something we'll need in this project